### ISSUE TITLE
ICE: `unexpected region: RePlaceholder(Placeholder` with mir drop tracking

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#110941]
### Code
`rustc treereduce.out --edition=2018 -Zdrop-tracking-mir=yes`
```Rust
use std::future::{Future, Ready};
async fn read() {}
async fn connect<A: ToSocketAddr>(addr: A) {
    let _ = addr.to_socket_addr().await;
}
pub trait ToSocketAddr {
    type Future: Future;
    fn to_socket_addr(&self) -> Self::Future;
}
impl ToSocketAddr for &() {
    type Future = Ready<()>;
    fn to_socket_addr(&self) -> Self::Future {
        unimplemented!()
    }
}
struct Server;
impl Server {
    fn and_then<F>(self, _fun: F) -> AndThen<F> {
        unimplemented!()
    }
}
struct AndThen<F> {
    _marker: std::marker::PhantomData<F>,
}
pub async fn run<F>(_: F) {
    let _ = connect(&()).await;
}
fn main() {
    let _ = async {
        let server = Server;
        let verification_route = server.and_then(read);
        run(verification_route).await;
    };
}

```

### Meta

`rustc --version --verbose`:
```
rustc 1.71.0-nightly (1c42cb4ef 2023-04-26)
binary: rustc
commit-hash: 1c42cb4ef0544fbfaa500216e53382d6b079c001
commit-date: 2023-04-26
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.2
```

### Error output

```
<output>
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_trait_selection/src/traits/select/mod.rs:3018:26: unexpected region: RePlaceholder(Placeholder { universe: U1, bound: BoundRegion { var: 0, kind: BrAnon(None) } })

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/compiler/rustc_errors/src/lib.rs:1643:9
stack backtrace:
   0:     0x7fa3211cfe73 - std::backtrace_rs::backtrace::libunwind::trace::h0c8179e89cc9893e
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7fa3211cfe73 - std::backtrace_rs::backtrace::trace_unsynchronized::hfadffaff9fceb6ba
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fa3211cfe73 - std::sys_common::backtrace::_print_fmt::ha49f609af156c4cb
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7fa3211cfe73 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h0ad104dc9ebb0bab
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7fa321230c8f - core::fmt::write::h076b507083203424
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/core/src/fmt/mod.rs:1247:17
   5:     0x7fa3211c2ea1 - std::io::Write::write_fmt::ha4c326e8c8c981e1
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/io/mod.rs:1712:15
   6:     0x7fa3211cfc85 - std::sys_common::backtrace::_print::hbc9e8503f2c8583b
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7fa3211cfc85 - std::sys_common::backtrace::print::he24fec1d9c5516a7
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7fa3211d2847 - std::panicking::default_hook::{{closure}}::hf29caaf6379226a6
   9:     0x7fa3211d2635 - std::panicking::default_hook::h0765c9476c979227
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/panicking.rs:288:9
  10:     0x7fa31ffbf8c5 - <rustc_driver_impl[b4ece89b41fa04d8]::DEFAULT_HOOK::{closure#0}::{closure#0} as core[a53159a38bb5d97c]::ops::function::FnOnce<(&core[a53159a38bb5d97c]::panic::panic_info::PanicInfo,)>>::call_once::{shim:vtable#0}
  11:     0x7fa3211d2f85 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h7a83df3f1b1b39ec
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/alloc/src/boxed.rs:1976:9
  12:     0x7fa3211d2f85 - std::panicking::rust_panic_with_hook::h060701a46b500fdf
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/panicking.rs:695:13
  13:     0x7fa3204cf0d1 - std[b0e36382822581c2]::panicking::begin_panic::<rustc_errors[e242ea0d30f7fc85]::ExplicitBug>::{closure#0}
  14:     0x7fa3204cdeb6 - std[b0e36382822581c2]::sys_common::backtrace::__rust_end_short_backtrace::<std[b0e36382822581c2]::panicking::begin_panic<rustc_errors[e242ea0d30f7fc85]::ExplicitBug>::{closure#0}, !>
  15:     0x7fa320495da6 - std[b0e36382822581c2]::panicking::begin_panic::<rustc_errors[e242ea0d30f7fc85]::ExplicitBug>
  16:     0x7fa3205174d6 - <rustc_errors[e242ea0d30f7fc85]::HandlerInner>::bug::<&alloc[5698eed4a099ff47]::string::String>
  17:     0x7fa3205171a0 - <rustc_errors[e242ea0d30f7fc85]::Handler>::bug::<&alloc[5698eed4a099ff47]::string::String>
  18:     0x7fa3205107bb - rustc_middle[74d31bc1c7288efd]::util::bug::opt_span_bug_fmt::<rustc_span[5acbb0d3e38bbf81]::span_encoding::Span>::{closure#0}
  19:     0x7fa32050e0fa - rustc_middle[74d31bc1c7288efd]::ty::context::tls::with_opt::<rustc_middle[74d31bc1c7288efd]::util::bug::opt_span_bug_fmt<rustc_span[5acbb0d3e38bbf81]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  20:     0x7fa32050e0ca - rustc_middle[74d31bc1c7288efd]::ty::context::tls::with_context_opt::<rustc_middle[74d31bc1c7288efd]::ty::context::tls::with_opt<rustc_middle[74d31bc1c7288efd]::util::bug::opt_span_bug_fmt<rustc_span[5acbb0d3e38bbf81]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  21:     0x7fa31e18cd0d - rustc_middle[74d31bc1c7288efd]::util::bug::bug_fmt
  22:     0x7fa320c22285 - rustc_trait_selection[ce8ecfa74af62748]::traits::select::bind_generator_hidden_types_above::{closure#0}::{closure#1}::{closure#0}
  23:     0x7fa320ba51c2 - <rustc_middle[74d31bc1c7288efd]::ty::Ty as rustc_type_ir[54773d718d3b4bd4]::fold::TypeSuperFoldable<rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt>>::super_fold_with::<rustc_middle[74d31bc1c7288efd]::ty::fold::RegionFolder>
  24:     0x7fa320c12cb9 - <rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt>::fold_regions::<rustc_middle[74d31bc1c7288efd]::ty::Ty, rustc_trait_selection[ce8ecfa74af62748]::traits::select::bind_generator_hidden_types_above::{closure#0}::{closure#1}::{closure#0}>
  25:     0x7fa320c22788 - <&mut rustc_trait_selection[ce8ecfa74af62748]::traits::select::bind_generator_hidden_types_above::{closure#0}::{closure#1} as core[a53159a38bb5d97c]::ops::function::FnOnce<(rustc_middle[74d31bc1c7288efd]::ty::subst::EarlyBinder<rustc_middle[74d31bc1c7288efd]::ty::Ty>,)>>::call_once
  26:     0x7fa320b5b94a - <alloc[5698eed4a099ff47]::vec::Vec<rustc_middle[74d31bc1c7288efd]::ty::Ty> as alloc[5698eed4a099ff47]::vec::spec_from_iter::SpecFromIter<rustc_middle[74d31bc1c7288efd]::ty::Ty, core[a53159a38bb5d97c]::iter::adapters::map::Map<core[a53159a38bb5d97c]::iter::adapters::filter::Filter<core[a53159a38bb5d97c]::iter::adapters::map::Map<core[a53159a38bb5d97c]::iter::adapters::filter::Filter<core[a53159a38bb5d97c]::slice::iter::Iter<rustc_middle[74d31bc1c7288efd]::mir::query::GeneratorSavedTy>, <rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt>::generator_hidden_types::{closure#0}>, <rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt>::generator_hidden_types::{closure#1}>, rustc_trait_selection[ce8ecfa74af62748]::traits::select::bind_generator_hidden_types_above::{closure#0}::{closure#0}>, rustc_trait_selection[ce8ecfa74af62748]::traits::select::bind_generator_hidden_types_above::{closure#0}::{closure#1}>>>::from_iter
  27:     0x7fa320c25514 - rustc_trait_selection[ce8ecfa74af62748]::traits::select::bind_generator_hidden_types_above
  28:     0x7fa31e8b02f8 - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::confirm_candidate
  29:     0x7fa31e8a0d6f - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_stack
  30:     0x7fa31e89f3f4 - <rustc_query_system[38efe1ae13b76606]::dep_graph::graph::DepGraph<rustc_middle[74d31bc1c7288efd]::dep_graph::dep_node::DepKind>>::with_anon_task::<rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt, <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::in_task<<rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_trait_predicate_recursively::{closure#0}::{closure#2}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>::{closure#0}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>
  31:     0x7fa31e89b02c - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_predicates_recursively::<alloc[5698eed4a099ff47]::vec::into_iter::IntoIter<rustc_infer[7992d7bd645369a3]::traits::Obligation<rustc_middle[74d31bc1c7288efd]::ty::Predicate>>>
  32:     0x7fa31e8a0dfe - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_stack
  33:     0x7fa31e89f3f4 - <rustc_query_system[38efe1ae13b76606]::dep_graph::graph::DepGraph<rustc_middle[74d31bc1c7288efd]::dep_graph::dep_node::DepKind>>::with_anon_task::<rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt, <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::in_task<<rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_trait_predicate_recursively::{closure#0}::{closure#2}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>::{closure#0}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>
  34:     0x7fa31e89b02c - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_predicates_recursively::<alloc[5698eed4a099ff47]::vec::into_iter::IntoIter<rustc_infer[7992d7bd645369a3]::traits::Obligation<rustc_middle[74d31bc1c7288efd]::ty::Predicate>>>
  35:     0x7fa31e8a0dfe - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_stack
  36:     0x7fa31e89f3f4 - <rustc_query_system[38efe1ae13b76606]::dep_graph::graph::DepGraph<rustc_middle[74d31bc1c7288efd]::dep_graph::dep_node::DepKind>>::with_anon_task::<rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt, <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::in_task<<rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_trait_predicate_recursively::{closure#0}::{closure#2}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>::{closure#0}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>
  37:     0x7fa31e89b02c - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_predicates_recursively::<alloc[5698eed4a099ff47]::vec::into_iter::IntoIter<rustc_infer[7992d7bd645369a3]::traits::Obligation<rustc_middle[74d31bc1c7288efd]::ty::Predicate>>>
  38:     0x7fa31e8a0dfe - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_stack
  39:     0x7fa31e89f3f4 - <rustc_query_system[38efe1ae13b76606]::dep_graph::graph::DepGraph<rustc_middle[74d31bc1c7288efd]::dep_graph::dep_node::DepKind>>::with_anon_task::<rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt, <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::in_task<<rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_trait_predicate_recursively::{closure#0}::{closure#2}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>::{closure#0}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>
  40:     0x7fa31e89b02c - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_predicates_recursively::<alloc[5698eed4a099ff47]::vec::into_iter::IntoIter<rustc_infer[7992d7bd645369a3]::traits::Obligation<rustc_middle[74d31bc1c7288efd]::ty::Predicate>>>
  41:     0x7fa31e8a0dfe - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_stack
  42:     0x7fa31e89f3f4 - <rustc_query_system[38efe1ae13b76606]::dep_graph::graph::DepGraph<rustc_middle[74d31bc1c7288efd]::dep_graph::dep_node::DepKind>>::with_anon_task::<rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt, <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::in_task<<rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_trait_predicate_recursively::{closure#0}::{closure#2}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>::{closure#0}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>
  43:     0x7fa31e89b02c - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_predicates_recursively::<alloc[5698eed4a099ff47]::vec::into_iter::IntoIter<rustc_infer[7992d7bd645369a3]::traits::Obligation<rustc_middle[74d31bc1c7288efd]::ty::Predicate>>>
  44:     0x7fa31e8a0dfe - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_stack
  45:     0x7fa31e89f3f4 - <rustc_query_system[38efe1ae13b76606]::dep_graph::graph::DepGraph<rustc_middle[74d31bc1c7288efd]::dep_graph::dep_node::DepKind>>::with_anon_task::<rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt, <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::in_task<<rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_trait_predicate_recursively::{closure#0}::{closure#2}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>::{closure#0}, core[a53159a38bb5d97c]::result::Result<rustc_middle[74d31bc1c7288efd]::traits::select::EvaluationResult, rustc_middle[74d31bc1c7288efd]::traits::select::OverflowError>>
  46:     0x7fa31eae2d2c - <rustc_trait_selection[ce8ecfa74af62748]::traits::select::SelectionContext>::evaluate_predicate_recursively
  47:     0x7fa31e67ae43 - rustc_traits[bbceeee60bcdb08d]::evaluate_obligation::evaluate_obligation
  48:     0x7fa31e678b6d - rustc_query_system[38efe1ae13b76606]::query::plumbing::try_execute_query::<rustc_query_impl[d894186be28c7657]::queries::evaluate_obligation, rustc_query_impl[d894186be28c7657]::plumbing::QueryCtxt>
  49:     0x7fa31ebbf3c8 - <rustc_infer[7992d7bd645369a3]::infer::InferCtxt as rustc_trait_selection[ce8ecfa74af62748]::traits::query::evaluate_obligation::InferCtxtExt>::evaluate_obligation_no_overflow
  50:     0x7fa31ec2122b - rustc_ty_utils[4df037ea12ea8272]::common_traits::is_item_raw
  51:     0x7fa31f351709 - rustc_query_system[38efe1ae13b76606]::query::plumbing::try_execute_query::<rustc_query_impl[d894186be28c7657]::queries::is_freeze_raw, rustc_query_impl[d894186be28c7657]::plumbing::QueryCtxt>
  52:     0x7fa31f351387 - <rustc_query_impl[d894186be28c7657]::Queries as rustc_middle[74d31bc1c7288efd]::ty::query::QueryEngine>::is_freeze_raw
  53:     0x7fa31ea776ad - <rustc_middle[74d31bc1c7288efd]::ty::Ty>::is_freeze
  54:     0x7fa31ea76aa9 - <rustc_mir_transform[913929990536c59]::ssa::SsaLocals>::new
  55:     0x7fa31ea71cf2 - <rustc_mir_transform[913929990536c59]::copy_prop::CopyProp as rustc_middle[74d31bc1c7288efd]::mir::MirPass>::run_pass
  56:     0x7fa31f358893 - rustc_mir_transform[913929990536c59]::optimized_mir
  57:     0x7fa31f356a20 - rustc_query_system[38efe1ae13b76606]::query::plumbing::try_execute_query::<rustc_query_impl[d894186be28c7657]::queries::optimized_mir, rustc_query_impl[d894186be28c7657]::plumbing::QueryCtxt>
  58:     0x7fa31edafe97 - <rustc_middle[74d31bc1c7288efd]::ty::context::TyCtxt>::instance_mir
  59:     0x7fa31ed9e70d - rustc_monomorphize[7be5ee009e525005]::collector::collect_neighbours
  60:     0x7fa31ed9aeca - rustc_monomorphize[7be5ee009e525005]::collector::collect_items_rec
  61:     0x7fa31f3f56cc - rustc_data_structures[ecde4e0aacf74356]::sync::par_for_each_in::<alloc[5698eed4a099ff47]::vec::Vec<rustc_middle[74d31bc1c7288efd]::mir::mono::MonoItem>, rustc_monomorphize[7be5ee009e525005]::collector::collect_crate_mono_items::{closure#1}::{closure#0}>
  62:     0x7fa31f3f5380 - <rustc_session[86cf650fda0448f5]::session::Session>::time::<(), rustc_monomorphize[7be5ee009e525005]::collector::collect_crate_mono_items::{closure#1}>
  63:     0x7fa31f3f4fa8 - rustc_monomorphize[7be5ee009e525005]::collector::collect_crate_mono_items
  64:     0x7fa31f3f3337 - rustc_monomorphize[7be5ee009e525005]::partitioning::collect_and_partition_mono_items
  65:     0x7fa31f7ecd94 - rustc_query_system[38efe1ae13b76606]::query::plumbing::try_execute_query::<rustc_query_impl[d894186be28c7657]::queries::collect_and_partition_mono_items, rustc_query_impl[d894186be28c7657]::plumbing::QueryCtxt>
  66:     0x7fa31f7eca5d - <rustc_query_impl[d894186be28c7657]::Queries as rustc_middle[74d31bc1c7288efd]::ty::query::QueryEngine>::collect_and_partition_mono_items
  67:     0x7fa31f513d75 - rustc_codegen_ssa[4ba1349a83ac51bf]::base::codegen_crate::<rustc_codegen_llvm[f1bb86015ace03e0]::LlvmCodegenBackend>
  68:     0x7fa31f513b39 - <rustc_codegen_llvm[f1bb86015ace03e0]::LlvmCodegenBackend as rustc_codegen_ssa[4ba1349a83ac51bf]::traits::backend::CodegenBackend>::codegen_crate
  69:     0x7fa31f1c1bb2 - <rustc_session[86cf650fda0448f5]::session::Session>::time::<alloc[5698eed4a099ff47]::boxed::Box<dyn core[a53159a38bb5d97c]::any::Any>, rustc_interface[69aa8c17999313ff]::passes::start_codegen::{closure#0}>
  70:     0x7fa31f1c16e0 - rustc_interface[69aa8c17999313ff]::passes::start_codegen
  71:     0x7fa31f1bc7b4 - <rustc_middle[74d31bc1c7288efd]::ty::context::GlobalCtxt>::enter::<<rustc_interface[69aa8c17999313ff]::queries::Queries>::ongoing_codegen::{closure#0}::{closure#0}, core[a53159a38bb5d97c]::result::Result<alloc[5698eed4a099ff47]::boxed::Box<dyn core[a53159a38bb5d97c]::any::Any>, rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>>
  72:     0x7fa31f1ba398 - <rustc_interface[69aa8c17999313ff]::queries::Queries>::ongoing_codegen
  73:     0x7fa31f1b9b7b - <rustc_interface[69aa8c17999313ff]::interface::Compiler>::enter::<rustc_driver_impl[b4ece89b41fa04d8]::run_compiler::{closure#1}::{closure#2}, core[a53159a38bb5d97c]::result::Result<core[a53159a38bb5d97c]::option::Option<rustc_interface[69aa8c17999313ff]::queries::Linker>, rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>>
  74:     0x7fa31f1b7b6f - rustc_span[5acbb0d3e38bbf81]::set_source_map::<core[a53159a38bb5d97c]::result::Result<(), rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>, rustc_interface[69aa8c17999313ff]::interface::run_compiler<core[a53159a38bb5d97c]::result::Result<(), rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>, rustc_driver_impl[b4ece89b41fa04d8]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  75:     0x7fa31f1b7200 - std[b0e36382822581c2]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[69aa8c17999313ff]::util::run_in_thread_pool_with_globals<rustc_interface[69aa8c17999313ff]::interface::run_compiler<core[a53159a38bb5d97c]::result::Result<(), rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>, rustc_driver_impl[b4ece89b41fa04d8]::run_compiler::{closure#1}>::{closure#0}, core[a53159a38bb5d97c]::result::Result<(), rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a53159a38bb5d97c]::result::Result<(), rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>>
  76:     0x7fa31f1b6b21 - <<std[b0e36382822581c2]::thread::Builder>::spawn_unchecked_<rustc_interface[69aa8c17999313ff]::util::run_in_thread_pool_with_globals<rustc_interface[69aa8c17999313ff]::interface::run_compiler<core[a53159a38bb5d97c]::result::Result<(), rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>, rustc_driver_impl[b4ece89b41fa04d8]::run_compiler::{closure#1}>::{closure#0}, core[a53159a38bb5d97c]::result::Result<(), rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a53159a38bb5d97c]::result::Result<(), rustc_span[5acbb0d3e38bbf81]::ErrorGuaranteed>>::{closure#1} as core[a53159a38bb5d97c]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  77:     0x7fa3211dd365 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h2b4cb02d9a218e1d
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/alloc/src/boxed.rs:1962:9
  78:     0x7fa3211dd365 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hbd8144eacfab6e10
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/alloc/src/boxed.rs:1962:9
  79:     0x7fa3211dd365 - std::sys::unix::thread::Thread::new::thread_start::h95499a4b2fca6f4a
                               at /rustc/1c42cb4ef0544fbfaa500216e53382d6b079c001/library/std/src/sys/unix/thread.rs:108:17
  80:     0x7fa31cc9ebb5 - <unknown>
  81:     0x7fa31cd20d90 - <unknown>
  82:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.71.0-nightly (1c42cb4ef 2023-04-26) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z drop-tracking-mir=yes

query stack during panic:
#0 [evaluate_obligation] evaluating trait selection obligation `[async block@treereduce.out:29:13: 33:6]: core::marker::Freeze`
#1 [is_freeze_raw] computing whether `[async block@treereduce.out:29:13: 33:6]` is freeze
#2 [optimized_mir] optimizing MIR for `main`
#3 [collect_and_partition_mono_items] collect_and_partition_mono_items
end of query stack
error: aborting due to previous error
```

</p>
</details>

**Comment by matthiaskrgr** [comment]
searched toolchains nightly-2023-04-26 through nightly-2023-04-28
Regression in nightly-2023-04-27
searched toolchains 458d4dae845ec155b285681a5b88305641abb868 through 1c42cb4ef0544fbfaa500216e53382d6b079c001
Regression in 1c42cb4ef0544fbfaa500216e53382d6b079c001
Attempting to search unrolled perf builds
Found commits ["de1f24d5", "8e20d409", "54fd5ffd", "50914bc9", "64cec96f", "5656a2d5", "42f2bef1", "f76c3084", "a9481ef7", "34bfa438"]
Regression in https://github.com/rust-lang-ci/rust/commit/8e20d40994195d763b18e32bf37cf204289a7eca

https://github.com/rust-lang/rust/pull/110835  cc @nnethercote

**Comment by nnethercote** [comment]
This was fixed by #107421.
